### ISSUE TITLE
[scale] node state changes causes excessive LB re-configurations 

### ISSUE THREAD

**Issue by alexanderConstantinescu** [kubernetes/kubernetes#111539]
### What happened?

There is scale bug related to the cloud-controller-manager, namely: when a cluster reaches a state where the relation between the amount of LBs and cluster nodes reaches an almost equal amount (i.e: a many-to-many relation between LBs and nodes) the amount of cloud API calls performed by the CCM to sync LBs can become overwhelming for the cloud provider. This can (among other things) lead to:

1. an excessive amount of cloud API calls for the LB re-configurations
2. rate limiting by the cloud provider as an effect of 1.
3. or worse: a non-trivial amount of time to sync all services in the cluster, leading to desynchronized LBs.

An example of problem 3. is: imagine we have 200 LBs and 500 nodes, on clusters at such scale, the amount of events processed by the CCM for transitioning node state can become important. The current implementation of the CCM will resync all 200 LB services for any node which has a transitioning state - as to keep the LBs' configured backends in sync with the state of the cluster. It has been observed that on clusters at that scale, the cloud provider can take up to hours to process a full resync for all services. A scenario such as the below is hence not improbable (and actually quite likely):

1. Node25 transitions from Ready -> NotReady
[All LB services are synced - time taken for this might be long]
2. Service1's LB which only has an endpoint on Node25 gets Node25 removed from its list of backend
3. Node25 transitions from NotReady -> Ready again, cancelling the scheduled eviction for the endpoint hosted on Node25 for Service1
4. CCM is still processing the updated caused by 1. and clients connecting to the LB of Service1 continues experiencing connectivity issues, even though Node25 now is in a Ready state.
[Eventually - once the CCM enters its second reconciliation loop caused by 3.]
5. The LB for Service1 gets Node25 added to its list of backend again and connectivity is again achieved by clients.

The sync function which is triggered when a node transitions state from Ready <-> NotReady, is currently correct but can be avoided for services which have ExternalTrafficPolicyLocal if the service does not have an endpoint on any of the nodes which are transitioning state between Ready <-> NotReady. This is because such services are not designed to forward traffic to any other node, and the update is thus useless to them. The node which is experiencing a transitioning state is already configured for that LB, and the fact that it transitions state is a moot point for what concerns these services. 

The longer vision I have for this: remove the sync function for transitioning node readiness state from the CCM altogether and instead have all services of type LoadBalancer evaluate the node's readiness state by the utilization of a LB configured health check probe against the [kubelet's read-only port](https://github.com/kubernetes/kubernetes/blob/master/pkg/cluster/ports/ports.go#L33). This would allow transitioning node readiness state to be handled much more dynamically (similar to what has already been implemented for HealthCheckNodePort by kube-proxy) for all services of type LoadBalancer irrespectively of if they define ExternalTrafficPolicyLocal or not....but that will require a enhancement proposal since cloud providers have divergent implementations of how they configure health check probes for LB services.

I am putting this here for completeness sake: this is what the major cloud providers currently do to health check services which are not ExternalTrafficPolicyLocal

GCE: probes [10256](https://github.com/kubernetes/kubernetes/blob/master/pkg/cluster/ports/ports.go#L36)
AWS: if ELB: probes the first NodePort defined on the service spec.
Azure: probes all NodePort defined on the service spec.

All of these can be improved, but an enhancement proposal would be needed to align them all. For now: this PR will help fix some cases.

### What did you expect to happen?

Node readiness changes should not re-configure load balancers associated with services of type ETP=local. 

- Nodes not hosting any endpoints associated with these services do not forward traffic to nodes that do, so re-configuring the LB because said node goes NotReady is useless. 
- Nodes hosting endpoints associated with these services are already health checked using the HC probe configured for the `healthCheckNodePort` and can rely on that much more dynamically without needing to re-configure the LB. 

In the long-term: all load balancers should not be re-configured because of node state changes. That will need an enhancement proposal however (which I am currently drafting) to change service proxies and align cloud provider HCs. 

### How can we reproduce it (as minimally and precisely as possible)?

"as minimally and precisely as possible" not sure if this hits the definition, but: we need a big cluster with a lot of entropy (for example by performing an upgrade and restarting all nodes one-by-one) - this should end up re-configuring every single LB N*M (where N is the amount of nodes and M is the amount of state transitions per node). 

### Anything else we need to know?

Presumably, this has never been a big issue for the externally exposed Kube API service, since that runs on control plane nodes...and I assume those nodes are labeled with https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/well_known_labels.go#L62-L67 (I am pretty sure that's the case for OpenShift). So "Kubernetes distros" avoid this problem using that label...users, don't get that luxury however.  

### Kubernetes version

All versions.

### Cloud provider

All cloud providers. 

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by alexanderConstantinescu** [comment]
I am doubtful I should open a new issue for this (seeing as how it's very much correlated to this issue), but there's another problem exherbitated by re-configuring the LB when not really needed; 

Updating a service from `ExternalTrafficPolicy = Cluster` to `ExternalTrafficPolicy = Local` can/will cause minute-wide blackouts for the service SLAs.

Specifically, when such an update is issued on the service object: kube-proxy will update all rules on nodes not hosting any endpoint and stop traffic from being forwarded, this is "the fast path". Inefficiently however, the CCM will also trigger a re-configuration of the LBs HC, which is the "slow path". Any client attempting to connect to the application backed by this service will not be able to until the LB is fully re-configured, since the LB won't have an HC or an incomplete one not being able to fully indicate where traffic should be sent. It feels ridiculous that such a small change can lead to such high windows of observed downtime - this is particularly impactful on low latency applications with strict SLAs.  

This can't be fixed today without that enhancement proposal, but I think that in an ideal world (hopefully once the KEP is in) the HC for what concerns the LB should remain the same. The only real change should on such an update would be the service proxy's response to that HC. That would severely reduce the observed "downtime". That obviously means impacting the `healthCheckNodePort`, since that port will need to be aligned with the default HC port for ETP!=Local services. At the same time users can specify what that port should be, so maybe the default port should change if the service has a `healthCheckNodePort` already defined? Te be discussed on the KEP...

**Comment by alexanderConstantinescu** [comment]
/reopen

This has not been fully completed by #109706, only partially.

**Comment by nckturner** [comment]
@alexanderConstantinescu do you think we need a KEP for the rest of this change?

**Comment by alexanderConstantinescu** [comment]
> @alexanderConstantinescu do you think we need a KEP for the rest of this change? (It sounds like there is already a KEP in progress, do you want to link it here?)

Sure, here's the KEP I posted and which hopefully will get reviewed during the 1.26 KEP cycle: https://github.com/kubernetes/enhancements/pull/3460

**Comment by elmiko** [comment]
marking this as accepted since there is a KEP in flight

 /triage accepted

**Comment by shaneutt** [comment]
@alexanderConstantinescu is this one resolved now? Or if not, what remains?

**Comment by alexanderConstantinescu** [comment]
> @alexanderConstantinescu is this one resolved now? Or if not, what remains?

Yes, this is the work covered and resolved by the KEP: https://github.com/kubernetes/enhancements/issues/3458 . This is fully resolved now.

**Comment by alvaroaleman** [comment]
/close

To not confuse people
